File: app/utils/trump.py
"""Trump rotation and round sequence utilities."""

TRUMP_ORDER = ["spades", "diamonds", "clubs", "hearts"]
ROUNDS_PER_SET = 8
DECK_SIZE = 52


def max_cards_for_players(player_count: int) -> int:
    """Max cards per round = floor(deck_size / players)."""
    return DECK_SIZE // player_count


def get_trump_for_round(round_num: int) -> str:
    """Get trump suit for a given round number (1-based)."""
    return TRUMP_ORDER[(round_num - 1) % len(TRUMP_ORDER)]


def get_cards_for_round(round_num: int, rounds_per_set: int = ROUNDS_PER_SET) -> int:
    """Get number of cards dealt for a given round.

    Pattern alternates: odd sets descend (8→1), even sets ascend (1→8).
    """
    position_in_set = (round_num - 1) % rounds_per_set
    set_number = (round_num - 1) // rounds_per_set  # 0-based
    if set_number % 2 == 0:
        # Odd sets (1st, 3rd, ...): descend 8→1
        return rounds_per_set - position_in_set
    else:
        # Even sets (2nd, 4th, ...): ascend 1→8
        return position_in_set + 1

File: app/static/js/engine/trump.js
/**
 * Trump rotation and round-card utilities.
 *
 * Port of app/utils/trump.py — exact parity with Python implementation.
 */

const DECK_SIZE = 52;
const TRUMP_ORDER = ["spades", "diamonds", "clubs", "hearts"];
const DEFAULT_ROUNDS_PER_SET = 8;

/**
 * Maximum cards that can be dealt per round given a player count.
 * Uses integer (floor) division of deck size.
 *
 * @param {number} playerCount
 * @returns {number}
 */
export function maxCardsForPlayers(playerCount) {
  return Math.floor(DECK_SIZE / playerCount);
}

/**
 * Get the trump suit name for a given round number (1-based).
 * Rotates: spades → diamonds → clubs → hearts, repeating.
 *
 * @param {number} roundNum  1-based round number
 * @returns {string}  suit name, e.g. "spades"
 */
export function getTrumpForRound(roundNum) {
  return TRUMP_ORDER[(roundNum - 1) % TRUMP_ORDER.length];
}

/**
 * Get number of cards dealt in a given round.
 * Pattern alternates by set: odd sets descend (8→1), even sets ascend (1→8).
 *
 * @param {number} roundNum        1-based round number
 * @param {number} roundsPerSet    cards in a full set (default 8)
 * @returns {number}
 */
export function getCardsForRound(roundNum, roundsPerSet = DEFAULT_ROUNDS_PER_SET) {
  const positionInSet = (roundNum - 1) % roundsPerSet;
  const setNumber = Math.floor((roundNum - 1) / roundsPerSet); // 0-based
  if (setNumber % 2 === 0) {
    // Odd sets (1st, 3rd, ...): descend roundsPerSet→1
    return roundsPerSet - positionInSet;
  } else {
    // Even sets (2nd, 4th, ...): ascend 1→roundsPerSet
    return positionInSet + 1;
  }
}
